Pinning workflow not detecting preexisting opam file in 1.2.1

I have an `opam` file in this new package, but then pin ignores it.

```
$ opam pin add lterm-progress-bar .
Package lterm-progress-bar does not exist, create as a NEW package ? [Y/n] y
lterm-progress-bar is now git-pinned to /Users/avsm/src/git/avsm/ocaml-progress-bar
[NOTE] Pinning in mixed mode: OPAM will use tracked files in the current working
       tree from /Users/avsm/src/git/avsm/ocaml-progress-bar. If this is not what you want, pin to a given branch (e.g.
       /Users/avsm/src/git/avsm/ocaml-progress-bar#HEAD)

[lterm-progress-bar] /Users/avsm/src/git/avsm/ocaml-progress-bar already up-to-date

Invalid opam file:
    error: File format error: at ~/.opam/system/overlay/lterm-progress-bar/opam_:9:11, Bad url format ""
Errors in /Users/avsm/.opam/system/overlay/lterm-progress-bar/opam, retry editing ? [Y/n] n
[ERROR] No valid metadata available.
lterm-progress-bar is now unpinned from git /Users/avsm/src/git/avsm/ocaml-progress-bar
```

It brings up an editor, despite there being an `opam` file in the current directory...

--- a/src/client/opamArg.ml
+++ b/src/client/opamArg.ml
@@ -1478,20 +1478,23 @@ let pin ?(unpin_only=false) () =
     ["list"]     , `list, [], "Lists pinned packages.";
     ["add"]      , `add  , ["PACKAGE"; "TARGET"],
     "Pins package $(i,PACKAGE) to $(i,TARGET), which may be a version, a path, \
-     or a url. \
+     or a URL. \
      $(i,PACKAGE) can be omitted if $(i,TARGET) is a local path containing a \
      package description with a name. $(i,TARGET) can be replaced by \
-     `--dev-repo'. \
-     OPAM will attempt to guess the desired pinning kind, unless you use \
-     URL syntax or an explicit $(b,--kind) option. \
-     Pins may target a specific branch or commit using $(b,#branch) e.g. \
-     $(b,git://host/me/pkg#testing). When they don't, in the special case \
-     of version-controlled pinning to a local path, OPAM will use \
-     \"mixed mode\": it will only use version-controlled files, but at their \
-     current version. \
-     It is possible to create a new package if $(i,NAME) does not exist. \
-     Its version may then be specified using $(i,NAME).$(i,VERSION), in the \
-     source opam file or with $(b,edit).";
+     `--dev-repo' if a package by that name is already known. \
+     OPAM will infer the kind of pinning from the format of $(i,TARGET), using \
+     $(b,path) pinning by default, unless you use an explicit $(b,--kind) \
+     option. \
+     Pins to version control systems may target a specific branch or commit \
+     using $(b,#branch) e.g. $(b,git://host/me/pkg#testing). When they don't, \
+     in the special case of version-controlled pinning to a local path, OPAM \
+     will use \"mixed mode\": it will only use version-controlled files, but \
+     at their current, on-disk version. \
+     If $(i,PACKAGE) is not a known package name, a new package by that name \
+     will be locally created. \
+     The package version may be specified by using the format \
+     $(i,NAME).$(i,VERSION) for $(PACKAGE), in the source opam file, or with \
+     $(b,edit).";
     ["remove"]   , `remove, ["NAMES"],
     "Unpins packages $(b,NAMES), restoring their definition from the \
      repository, if any.";
@@ -1525,8 +1528,14 @@ let pin ?(unpin_only=false) () =
     mk_flag ["e";"edit"] "With $(opam pin add), edit the opam file as with \
                           `opam pin edit' after pinning." in
   let kind =
-    let help = "Set the kind of pinning. Must be one of version, \
-                path, git, hg or darcs." in
+    let help =
+      "Sets the kind of pinning. Must be one of $(i,version), $(i,path), \
+       $(i,git), $(i,hg), $(i,darcs) or $(i,auto). \
+       If unset, is inferred from the format of the target, defaulting \
+       to $(i,path). If $(i,auto) or $(i,OPAMPINKINDAUTO) is set, a local \
+       path will be searched for version control and the pinning kind set \
+       accordingly. This is expected to become the default in a next version."
+    in
     let doc = Arg.info ~docv:"KIND" ~doc:help ["k";"kind"] in
     let kinds = [
       "http"   , `http;
@@ -1536,7 +1545,8 @@ let pin ?(unpin_only=false) () =
       "path"   , `local;
       "local"  , `local;
       "rsync"  , `local;
-      "hg"     , `hg
+      "hg"     , `hg;
+      "auto"   , `auto;
     ] in
     Arg.(value & opt (some & enum kinds) None & doc) in
   let no_act =
@@ -1574,6 +1584,10 @@ let pin ?(unpin_only=false) () =
   let pin global_options kind edit no_act dev_repo print_short command params =
     apply_global_options global_options;
     let action = not no_act in
+    let kind, guess = match kind with
+      | Some `auto -> None, true
+      | Some (#pin_kind as k) -> Some k, false
+      | None -> None, !OpamGlobals.pin_kind_auto in
     match command, params with
     | Some `list, [] | None, [] -> `Ok (Client.PIN.list ~short:print_short ())
     | Some `remove, names ->
@@ -1598,7 +1612,7 @@ let pin ?(unpin_only=false) () =
     | Some `add, [path] when not dev_repo ->
       (try
          let name = guess_name (OpamFilename.Dir.of_string path) in
-         let pin_option = pin_option_of_string ?kind path in
+         let pin_option = pin_option_of_string ?kind ~guess path in
          `Ok (Client.PIN.pin name ~edit ~action (Some pin_option))
        with Not_found ->
         `Error (false, Printf.sprintf
@@ -1609,7 +1623,7 @@ let pin ?(unpin_only=false) () =
     | Some `add, [n; target] ->
       (match (fst package) n with
        | `Ok (name,version) ->
-         let pin_option = pin_option_of_string ?kind:kind target in
+         let pin_option = pin_option_of_string ?kind ~guess target in
          `Ok (Client.PIN.pin name ?version ~edit ~action (Some pin_option))
        | `Error e -> `Error (false, e))
     | command, params -> bad_subcommand "pin" commands command params

--- a/src/client/opamArg.ml
+++ b/src/client/opamArg.ml
@@ -196,6 +196,8 @@ let apply_build_options b =
   | None   -> ()
   | Some s -> OpamGlobals.makecmd := (fun () -> s)
 
+let when_enum = [ "always", `Always; "never", `Never; "auto", `Auto ]
+
 (* Help sections common to all commands *)
 let global_option_section = "COMMON OPTIONS"
 let help_sections = [
@@ -224,6 +226,8 @@ let help_sections = [
   `P "$(i,OPAMNO) answer no to any question asked.";
   `P "$(i,OPAMNOASPCUD) see option `--no-aspcud'.";
   `P "$(i,OPAMNOSELFUPGRADE) see option `--no-self-upgrade'.";
+  `P "$(i,OPAMPINKINDAUTO) if set, version control systems are detected when \
+      pinning to a local path.";
   `P "$(i,OPAMREQUIRECHECKSUMS) see option `--require-checksums'.";
   `P "$(i,OPAMRETRY) sets the number of tries before failing downloads.";
   `P "$(i,OPAMROOT) see option `--root'. This is automatically set by \
@@ -233,9 +237,9 @@ let help_sections = [
   `P "$(i,OPAMSKIPVERSIONCHECKS) bypasses some version checks. Unsafe, for \
       compatibility testing only.";
   `P "$(i,OPAMSOLVERTIMEOUT) change the time allowance of the internal solver.";
-  `P "$(i,OPAMSTATUSLINE) display a dynamic status line showing what's \
-      currently going on on the terminal. \
-      (one of $(i,always), $(i,never) or $(i,auto))";
+  `P ("$(i,OPAMSTATUSLINE) display a dynamic status line showing what's \
+       currently going on on the terminal. \
+       (one of "^Arg.doc_alts_enum when_enum^")");
   `P "$(i,OPAMSWITCH) see option `--switch'. Automatically set by \
       `opam config env --switch=SWITCH'.";
   `P ("$(i,OPAMUPGRADECRITERIA) specifies user $(i,preferences) for dependency \
@@ -244,9 +248,9 @@ let help_sections = [
        The default value is "^OpamGlobals.default_preferences `Upgrade^
       ". See also option --criteria");
   `P "$(i,OPAMUSEINTERNALSOLVER) see option `--use-internal-solver'.";
-  `P "$(i,OPAMUTF8) use UTF8 characters in output \
-      (one of $(i,always), $(i,never) or $(i,auto), the default, which is \
-      determined from the locale).";
+  `P ("$(i,OPAMUTF8) use UTF8 characters in output \
+       (one of "^Arg.doc_alts_enum when_enum^
+      "). By default `auto', which is determined from the locale).");
   `P "$(i,OPAMUTF8MSGS) use extended UTF8 characters (camels) in OPAM \
       messages. Implies $(i,OPAMUTF8). This is set by default on OSX only.";
   `P "$(i,OPAMVAR_var) overrides the contents of the variable $(i,var)  when \
@@ -373,9 +377,7 @@ let mk_opt ?section ?vopt flags value doc conv default =
 
 let mk_tristate_opt ?section flags value doc auto default =
   let doc = Arg.info ?docs:section ~docv:value ~doc flags in
-  let choices =
-    Arg.enum [ "always", `Always; "never", `Never; "auto", `Auto ] in
-  let arg = Arg.(value & opt choices default & doc) in
+  let arg = Arg.(value & opt (enum when_enum) default & doc) in
   let to_bool = function
     | `Always -> true
     | `Never -> false
@@ -469,8 +471,7 @@ let installed_roots_flag =
   mk_flag ["installed-roots"] "Display only the installed roots."
 
 let shell_opt =
-  let enum =
-    Arg.enum [
+  let enum = [
       "bash",`bash;
       "sh",`sh;
       "csh",`csh;
@@ -478,9 +479,10 @@ let shell_opt =
       "fish",`fish;
     ] in
   mk_opt ["shell"] "SHELL"
-    "Sets the configuration mode for OPAM environment appropriate for $(docv). \
-     One of `bash', `sh', `csh', `zsh' or `fish'."
-    enum (OpamMisc.guess_shell_compat ())
+    (Printf.sprintf
+       "Sets the configuration mode for OPAM environment appropriate for \
+        $(docv). One of %s." (Arg.doc_alts_enum enum))
+    (Arg.enum enum) (OpamMisc.guess_shell_compat ())
 
 let dot_profile_flag =
   mk_opt ["dot-profile"]
@@ -489,23 +491,22 @@ let dot_profile_flag =
     (Arg.some filename) None
 
 let repo_kind_flag =
-  let kinds = [
-    (* main kinds *)
+  let main_kinds = [
     "http" , `http;
     "local", `local;
     "git"  , `git;
-    "darcs"  , `darcs;
+    "darcs", `darcs;
     "hg"   , `hg;
-
-    (* aliases *)
+  ] in
+  let aliases_kinds = [
     "wget" , `http;
     "curl" , `http;
     "rsync", `local;
   ] in
-  mk_opt ["k";"kind"]
-    "KIND" "Specify the kind of the repository to be set (the main ones \
-            are 'http', 'local', 'git', 'darcs' or 'hg')."
-    Arg.(some (enum kinds)) None
+  mk_opt ["k";"kind"] "KIND"
+    (Printf.sprintf "Specify the kind of the repository to be used (%s)."
+       (Arg.doc_alts_enum main_kinds))
+    Arg.(some (enum (main_kinds @ aliases_kinds))) None
 
 let jobs_flag =
   mk_opt ["j";"jobs"] "JOBS"
@@ -561,7 +562,8 @@ let global_options =
     mk_flag ~section ["q";"quiet"] "Be quiet when installing a new compiler." in
   let color =
     mk_tristate_opt ~section ["color"] "WHEN"
-      "Colorize the output. $(docv) must be `always', `never' or `auto'."
+      (Printf.sprintf "Colorize the output. $(docv) must be %s."
+         (Arg.doc_alts_enum when_enum))
       (fun () -> Unix.isatty Unix.stdout) OpamGlobals.color_when in
   let switch =
     mk_opt ~section ["switch"]
@@ -1528,26 +1530,29 @@ let pin ?(unpin_only=false) () =
     mk_flag ["e";"edit"] "With $(opam pin add), edit the opam file as with \
                           `opam pin edit' after pinning." in
   let kind =
-    let help =
-      "Sets the kind of pinning. Must be one of $(i,version), $(i,path), \
-       $(i,git), $(i,hg), $(i,darcs) or $(i,auto). \
-       If unset, is inferred from the format of the target, defaulting \
-       to $(i,path). If $(i,auto) or $(i,OPAMPINKINDAUTO) is set, a local \
-       path will be searched for version control and the pinning kind set \
-       accordingly. This is expected to become the default in a next version."
-    in
-    let doc = Arg.info ~docv:"KIND" ~doc:help ["k";"kind"] in
-    let kinds = [
+    let main_kinds = [
+      "version", `version;
+      "path"   , `local;
       "http"   , `http;
       "git"    , `git;
       "darcs"  , `darcs;
-      "version", `version;
-      "path"   , `local;
-      "local"  , `local;
-      "rsync"  , `local;
       "hg"     , `hg;
       "auto"   , `auto;
     ] in
+    let help =
+      Printf.sprintf
+        "Sets the kind of pinning. Must be one of %s. \
+         If unset, is inferred from the format of the target, defaulting \
+         to $(i,path). If $(i,auto) or $(i,OPAMPINKINDAUTO) is set, a local \
+         path will be searched for version control and the pinning kind set \
+         accordingly. This is expected to become the default in a next version."
+        (Arg.doc_alts_enum main_kinds)
+    in
+    let doc = Arg.info ~docv:"KIND" ~doc:help ["k";"kind"] in
+    let kinds = main_kinds @ [
+        "local"  , `local;
+        "rsync"  , `local;
+      ] in
     Arg.(value & opt (some & enum kinds) None & doc) in
   let no_act =
     mk_flag ["n";"no-action"]

--- a/src/client/opamSolution.ml
+++ b/src/client/opamSolution.ml
@@ -574,7 +574,9 @@ let apply ?ask t action ~requested solution =
       OpamGlobals.msg
         "The following actions %s be %s:\n"
         (if !OpamGlobals.show then "would" else "will")
-        (if !OpamGlobals.fake then "simulated" else "performed");
+        (if !OpamGlobals.dryrun then "simulated" else
+         if !OpamGlobals.fake then "faked"
+         else "performed");
       let new_state = simulate_new_state t action_graph in
       let messages p =
         let opam = OpamState.opam new_state p in

--- a/src/core/opamGlobals.ml
+++ b/src/core/opamGlobals.ml
@@ -92,6 +92,7 @@ let safe_mode        = check "SAFE"
 let lock_retries     =
   try ref (int_of_string (OpamMisc.getenv ("OPAMLOCKRETRIES")))
   with Not_found | Failure _ -> ref 5
+let pin_kind_auto    = check "PINKINDAUTO"
 let all_parens       = ref false
 
 (* Value set when opam calls itself *)

--- a/src/core/opamGlobals.mli
+++ b/src/core/opamGlobals.mli
@@ -47,6 +47,7 @@ val no_self_upgrade : bool ref
 val skip_version_checks : bool ref
 val safe_mode : bool ref
 val lock_retries : int ref
+val pin_kind_auto : bool ref
 
 (** Tells the printer to write parens everywhere *)
 val all_parens : bool ref

--- a/src/core/opamTypesBase.ml
+++ b/src/core/opamTypesBase.ml
@@ -135,7 +135,7 @@ let pin_of_url (url,kind) = match kind with
   | `hg -> Hg url
   | `local | `version -> failwith "Not a recognised version-control URL"
 
-let pin_option_of_string ?kind s =
+let pin_option_of_string ?kind ?(guess=false) s =
   match kind with
   | Some `version -> Version (OpamPackage.Version.of_string s)
   | None | Some (`http|`git|`hg|`darcs|`local) ->
@@ -155,9 +155,10 @@ let pin_option_of_string ?kind s =
       Local (OpamFilename.Dir.of_string (fst s))
     | None, `local ->
       let dir = OpamFilename.Dir.of_string (fst s) in
-      match guess_version_control dir with
-      | Some vc -> pin_of_url (s, vc)
-      | None -> Local dir
+      if guess then match guess_version_control dir with
+        | Some vc -> pin_of_url (s, vc)
+        | None -> Local dir
+      else Local dir
 
 let string_of_pin_kind = function
   | `version -> "version"

--- a/src/core/opamTypesBase.mli
+++ b/src/core/opamTypesBase.mli
@@ -76,8 +76,10 @@ val pin_kind_of_string: string -> pin_kind
 (** Parsing of pin kinds *)
 val string_of_pin_kind: pin_kind -> string
 
-(** Read pin options args *)
-val pin_option_of_string: ?kind:pin_kind -> string -> pin_option
+(** Read pin options args. If [kind] isn't specified, [guess] is set to [true]
+    and the name isn't explicit, look for VC on the filesystem to get the
+    pinning kind *)
+val pin_option_of_string: ?kind:pin_kind -> ?guess:bool -> string -> pin_option
 
 (** Convert a pin option to a string *)
 val string_of_pin_option: pin_option -> string

--- a/src_ext/Makefile
+++ b/src_ext/Makefile
@@ -8,8 +8,8 @@ MD5_extlib = 3de5f4e0a95fda7b2f3819c4a655b17c
 URL_re = https://github.com/ocaml/ocaml-re/archive/ocaml-re-1.2.0.tar.gz
 MD5_re = 5cbfc137683ef2b0e91f931577f2e673
 
-URL_cmdliner = http://erratique.ch/software/cmdliner/releases/cmdliner-0.9.4.tbz
-MD5_cmdliner = 5089b4b69993ddc5ce4aac0e30d0c641
+URL_cmdliner = http://erratique.ch/software/cmdliner/releases/cmdliner-0.9.7.tbz
+MD5_cmdliner = 46d7553b95f623f811df2a844fb448fb
 
 URL_graph = http://ocamlgraph.lri.fr/download/ocamlgraph-1.8.5.tar.gz
 MD5_graph = 75dde65bfc3f9b07e795343d369aa84d

--- a/src_ext/patches/cmdliner/backport_pre_4_00_0.patch
+++ b/src_ext/patches/cmdliner/backport_pre_4_00_0.patch
@@ -1,0 +1,30 @@
+diff -Naur cmdliner-0.9.7/src/cmdliner.ml cmdliner-0.9.7.patched/src/cmdliner.ml
+--- cmdliner-0.9.7/src/cmdliner.ml	2015-02-06 11:33:44.000000000 +0100
++++ cmdliner-0.9.7.patched/src/cmdliner.ml	2015-02-18 23:04:04.000000000 +0100
+@@ -830,7 +830,7 @@
+     | None -> if names = [] then "ARGUMENTS" else "OPTIONS"
+     | Some s -> s
+     in
+-    { id = arg_id (); absent = Val (Lazy.from_val "");
++    { id = arg_id (); absent = Val (lazy "");
+       doc = doc; docv = docv; docs = docs;
+       p_kind = All; o_kind = Flag; o_names = List.rev_map dash names;
+       o_all = false; }
+@@ -924,7 +924,7 @@
+ 
+   let opt_all ?vopt (parse, print) v a =
+     if is_pos a then invalid_arg err_not_opt else
+-    let a = { a with absent = Val (Lazy.from_val ""); o_all = true;
++    let a = { a with absent = Val (lazy ""); o_all = true;
+                      o_kind = match vopt with
+                      | None -> Opt | Some dv -> Opt_vopt (str_of_pp print dv) }
+     in
+@@ -949,7 +949,7 @@
+   let pos ?(rev = false) k (parse, print) v a =
+     if is_opt a then invalid_arg err_not_pos else
+     let a = { a with p_kind = Nth (rev, k);
+-                     absent = Val (Lazy.from_val (str_of_pp print v)) }
++                     absent = Val (lazy (str_of_pp print v)) }
+     in
+     let convert _ cl = match Cmdline.pos_arg cl a with
+     | [] -> v

